### ISSUE TITLE
Underscore('_') removal too arbitrary in _safe_attr

### ISSUE THREAD

**Issue by pwwang** [cdgriffith/Box#87]
Consider this code:
```python
from box import Box
b = Box(_out = 'preserved')
b.update({'out': 'updated'})
# expected:
# {'_out': 'preserved', 'out': 'updated'}
# observed:
# {'_out': 'updated'}
# out lost
```
I thought this was designed intentionally at the beginning. However, when I digged into the code, I found the problem was due to https://github.com/cdgriffith/Box/blob/master/box.py#L151
```python
    out = ''
    for character in attr:
        out += character if character in allowed else "_"
    out = out.strip("_")  <- L151
```
This arbitrarily removed all underscores in the key, even the originally assigned ones.  
I won't say this is a bug, but really confusing.  

Maybe we could use some unique characters to replace the characters that are not allowed, which will be removed later.

Say:  
```python
    out = ''
    for character in attr:
        out += character if character in allowed else "@"
    out = out.strip("@")
```

**Comment by cdgriffith** [comment]
Thanks for pointing this out! 

I am actually going to modify the logic a bit so it is safer all around

```python 
    out = ''
    starting_chr = ''
    for character in attr:
        if starting_chr:
            out += character if character in allowed else "_"
        elif character in allowed:
            starting_chr = character
            out += character

    out = out.rstrip("_")
```

**Comment by pwwang** [comment]
Will this overwrite the keys like `out_` by `out`?

**Comment by cdgriffith** [comment]
Gah, good point. Need to fix it from that side as well.

**Comment by cdgriffith** [comment]
After thinking about it some more, this is going to have to hold off until a major release. It is a bug fix, however any exiting code that uses entities like these would break on update. I am actively working on 4.0 and will slate it for that.

**Comment by cdgriffith** [comment]
Fixed in 4.0
